### PR TITLE
Implement dateadd macro DNA-15944 [DNA-20679]

### PR DESCRIPTION
**Pull request by silviustanimir**
## Description
- Implement charindex macro for Spark compatbility.

## Release
- [x] Direct release (`main`)
- [ ] Merge to `dev` (or other) branch
  - Why:

### Did you consider?
- [x] Does it Work on Automation Suite / SQL Server
- [x] Does it Work on Automation Cloud / Snowflake
- [ ] ~~What is the performance impact?~~
- [ ] ~~The version number in `dbt_project.yml`~~

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -79,6 +80,12 @@ In case you want to create the index on a source table, refer to the table using
 ) }}
 ```
 
+#### dateadd ([source](macros/multiple_databases/dateadd.sql))
+This macro adds a specified number value (as a signed integer) to a specified datepart of an input date value, and then returns that modified value. The datepart can be any of the following values for SQL Server and Snowflake: year, quarter, month, week, day, hour, minute, second, millisecond. For Databricks, the datepart can be any of the following values: year, day, hour, minute, second.
```
Comment (Dennis-UiPath) [inline README.md:84]: We are expecting a date or a datetime as input for this function, right? Can this be mentioned here explicitly. Now it states "input date value" and in the example uses "expression" (which can be a string)

File: macros/multiple_databases/dateadd.sql [hunk 2]
```diff
@@ -0,0 +1,22 @@
+{%- macro dateadd(datepart, number, datetime_field) -%}
+{%- if target.type == 'snowflake' -%}
+    dateadd({{ datepart }}, {{ number }}, try_to_timestamp(to_varchar({{ datetime_field }})))
```
Comment (Dennis-UiPath) [inline macros/multiple_databases/dateadd.sql:3]: If we know the datetime_field is a date or a datetime, we can use to_timestamp({{datetime_field}})

File: macros/multiple_databases/dateadd.sql [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+{%- elif target.type == 'sqlserver' -%}
+    dateadd({{ datepart }}, {{ number }}, try_convert(datetime2, {{ datetime_field }}))
+{%- elif target.type == 'databricks' -%}
+    {%- set clock_component -%}
```
Comment (Dennis-UiPath) [inline macros/multiple_databases/dateadd.sql:7]: can we rename this value to time_value? It represents the time only part of the datetime

File: macros/multiple_databases/dateadd.sql [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+        to_unix_timestamp({{ datetime_field }}) - to_unix_timestamp(date_trunc('DD', {{ datetime_field }}))
```
Comment (Dennis-UiPath) [inline macros/multiple_databases/dateadd.sql:8]: If we convert the input value using to_unix_timestamp, we will lose the millisecond precision. This makes it different from the Snowflake and SQL Server implementations

**Review by Dennis-UiPath — CHANGES_REQUESTED** [review]

File: README.md [hunk 5]
```diff
@@ -86,6 +87,12 @@ In case you want to create the index on a source table, refer to the table using
 ) }}
 ```
 
+#### dateadd ([source](macros/multiple_databases/dateadd.sql))
+This macro adds a specified number value (as a signed integer) to a specified datepart of an input date or datetime, and then returns that modified value. The datepart can be any of the following values for SQL Server and Snowflake: year, quarter, month, week, day, hour, minute, second, millisecond. For Databricks, the datepart can be any of the following values: year, quarter, month, week, day, hour, minute, second, millisecond.
```
[the remark this answers is not in the document]
Comment (silviustanimir) [reply README.md:91]: Actually there are more for [SQL Server](https://learn.microsoft.com/en-us/sql/t-sql/functions/dateadd-transact-sql?view=sql-server-ver16), but I will remove the second list.

**Review by Dennis-UiPath — APPROVED** [review]

**Pull request merged**